[1.3.236.0] CMake: No rule to make target 'loader/gen_defines.asm'

**Describe the bug**
Build of 1.3.236.0 may fail with `make[2]: *** No rule to make target 'loader/gen_defines.asm', needed by 'loader/CMakeFiles/loader_asm_gen_files'.  Stop.`

**Environment (please complete the following information):**
 - OS: Ubuntu
 - Bitdepth: 64-bit
 - SDK or header version if building from repo: vulkan-headers 1.3.236.0

**To Reproduce**
Build 1.3.236.0 with CMake (3.15 but I'm not sure), Unix Makefiles generator, gcc 11.

**Additional context**

I've submitted Vulkan-Loader 1.3.236.0 to conan-center, and I have no control other images used for the build. Several builds of 1.3.236.0 were successful on Linux, but at least one failed (with gcc 11), so there might be some issue related to CMake version.
All builds here: https://c3i.jfrog.io/c3i/misc/summary.html?json=https://c3i.jfrog.io/c3i/misc/logs/pr/14739/1-linux-gcc/vulkan-loader/1.3.236.0//summary.json
Failure: https://c3i.jfrog.io/c3i/misc/logs/pr/14739/1-linux-gcc/vulkan-loader/1.3.236.0//e5b55e1026379d04893834edbfeb89d0b36ad6fa-build.txt

I see that there were several changes since the last sdk version related to generation of `gen_defines.asm` and its add_custom_command using now `POST_BUILD` and `BYPRODUCTS` instead of `add_custom_command(OUTPUT gen_defines.asm DEPENDS asm_offset COMMAND asm_offset GAS)`: https://github.com/KhronosGroup/Vulkan-Loader/pull/1019 then https://github.com/KhronosGroup/Vulkan-Loader/pull/1039, so maybe something to investigate around these commands.

I can't test quickly (it's not my machine, and each time I push a change it builds a matrix of dozens of vulkan-loader versions with 130 different configurations).
But not need to check these variables, `ASSEMBLER_WORKS` is TRUE, otherwise `asm-offset` executable wouldn't be built: 

```
[  5%] Building C object loader/CMakeFiles/asm_offset.dir/asm_offset.c.o
[ 10%] Linking C executable asm_offset
[ 10%] Built target asm_offset
Makefile:129: recipe for target 'all' failed
make[2]: *** No rule to make target 'loader/gen_defines.asm', needed by 'loader/CMakeFiles/loader_asm_gen_files'.  Stop.
make[1]: *** [loader/CMakeFiles/loader_asm_gen_files.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
make: *** [all] Error 2
```

And for `CMAKE_CROSSCOMPILING`, it's FALSE for sure.

I think I understand: it comes from `BYPRODUCTS`, this change requires at least CMake 3.16 to ensure proper behavior of `add_custom_target(loader_asm_gen_files DEPENDS gen_defines.asm)`: https://cmake.org/cmake/help/latest/command/add_custom_target.html

```
Changed in version 3.16: A target-level dependency is added if any dependency is a byproduct of a target or any of its build events in the same directory to ensure the byproducts will be available before this target is built.
```

conan-center may use CMake 3.15 in some agents, so it fails.

--- a/loader/CMakeLists.txt
+++ b/loader/CMakeLists.txt
@@ -157,9 +157,7 @@ if(WIN32)
         target_link_libraries(asm_offset PRIVATE loader_specific_options)
         # If not cross compiling, run asm_offset to generage gen_defines.asm
         if (NOT CMAKE_CROSSCOMPILING)
-            add_custom_command(TARGET asm_offset POST_BUILD
-                COMMAND asm_offset MASM
-                BYPRODUCTS gen_defines.asm)
+            add_custom_command(OUTPUT gen_defines.asm DEPENDS asm_offset COMMAND asm_offset MASM)
         else()
             # Forces compiler to write the intermediate asm file, needed so that we can get sizeof/offset of info out of it.
             target_compile_options(asm_offset PRIVATE "/Fa$<TARGET_FILE_DIR:asm_offset>/asm_offset.asm" /FA)
@@ -233,9 +231,7 @@ else() # i.e.: Linux
         target_link_libraries(asm_offset loader_specific_options)
         # If not cross compiling, run asm_offset to generage gen_defines.asm
         if (NOT CMAKE_CROSSCOMPILING)
-            add_custom_command(TARGET asm_offset POST_BUILD
-                COMMAND asm_offset GAS
-                BYPRODUCTS gen_defines.asm)
+            add_custom_command(OUTPUT gen_defines.asm DEPENDS asm_offset COMMAND asm_offset GAS)
         else()
             # Forces compiler to write the intermediate asm file, needed so that we can get sizeof/offset of info out of it.
             target_compile_options(asm_offset PRIVATE -save-temps=obj)

